File: backend/routes/assistant.py
from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from starlette.concurrency import run_in_threadpool

from data.assistant_knowledge import SOURCES
from services.farming_assistant import (
    AssistantConfigError,
    AssistantProviderError,
    farming_assistant,
)


router = APIRouter(prefix="/assistant", tags=["farming assistant"])


class ConversationMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=2000)


class AssistantRequest(BaseModel):
    question: str = Field(min_length=2, max_length=1200)
    language: Literal["en", "si", "ta"] = "en"
    history: list[ConversationMessage] = Field(default_factory=list, max_length=8)


@router.get("/sources")
def list_sources() -> dict:
    return {"sources": [{"id": source_id, **source} for source_id, source in SOURCES.items()]}


@router.post("/chat")
async def chat(payload: AssistantRequest) -> dict:
    question = payload.question.strip()
    if len(question) < 2:
        raise HTTPException(status_code=422, detail="Enter a question about rice health or field management.")
    try:
        return await run_in_threadpool(
            farming_assistant.ask,
            question,
            payload.language,
            [message.model_dump() for message in payload.history],
        )
    except AssistantConfigError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    except AssistantProviderError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc

File: frontend/src/services/assistant.js
import axios from 'axios'
import api from './api'

function assistantError(error) {
  if (axios.isAxiosError(error) && error.response?.data?.detail) {
    return new Error(error.response.data.detail)
  }
  return new Error('The farming assistant is temporarily unavailable.')
}

export async function askFarmingAssistant(question, language, history) {
  try {
    const response = await api.post('/assistant/chat', { question, language, history })
    return response.data
  } catch (error) {
    throw assistantError(error)
  }
}

export async function getAssistantSources() {
  try {
    const response = await api.get('/assistant/sources')
    return response.data.sources
  } catch (error) {
    throw assistantError(error)
  }
}
